File: PathoFroh/src/main/java/com/patho/main/repository/service/impl/TaskRepositoryImpl.java
package com.patho.main.repository.service.impl;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Expression;
import javax.persistence.criteria.Join;
import javax.persistence.criteria.JoinType;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;
import javax.persistence.criteria.Subquery;

import com.patho.main.model.patient.*;
import com.patho.main.model.patient.notification.ReportIntent;
import org.hibernate.Hibernate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.patho.main.common.ContactRole;
import com.patho.main.common.Eye;
import com.patho.main.model.Person;
import com.patho.main.model.Physician;
import com.patho.main.model.Signature;
import com.patho.main.model.StainingPrototype;
import com.patho.main.repository.service.TaskRepositoryCustom;
import com.patho.main.util.helper.HistoUtil;
import com.patho.main.util.helper.TimeUtil;
import com.patho.main.util.worklist.search.WorklistSearchExtended;

@Service
@Transactional
public class TaskRepositoryImpl extends AbstractRepositoryCustom implements TaskRepositoryCustom {

	public Optional<Task> findOptionalByIdAndInitialize(Long id, boolean loadCouncils, boolean loadDiangoses,
			boolean loadPDFs, boolean loadContacts, boolean loadParent) {

		CriteriaBuilder qb = getCriteriaBuilder();
		CriteriaQuery<Task> criteria = qb.createQuery(Task.class);
		Root<Task> root = criteria.from(Task.class);

		return find(criteria, root, Arrays.asList(qb.equal(root.get(Task_.id), id)), loadCouncils, loadDiangoses,
				loadPDFs, loadContacts, loadParent);
	}

	public Optional<Task> findOptionalByTaskId(String taskID) {
		return findOptionalByTaskId(taskID, false, false, false, false, false);
	}

	public Optional<Task> findOptionalByTaskId(String taskID, boolean loadCouncils, boolean loadDiangoses,
			boolean loadPDFs, boolean loadContacts, boolean loadParent) {
		CriteriaBuilder qb = getSession().getCriteriaBuilder();

		// Create CriteriaQuery
		CriteriaQuery<Task> criteria = qb.createQuery(Task.class);
		Root<Task> root = criteria.from(Task.class);
		criteria.select(root);

		return find(criteria, root, Arrays.asList(qb.like(root.get("taskID"), taskID)), loadCouncils, loadDiangoses,
				loadPDFs, loadContacts, loadParent);
	}

	public Optional<Task> findOptionalBySlideID(String taskID, int uniqueSlideIDInBlock, boolean loadCouncils,
			boolean loadDiangoses, boolean loadPDFs, boolean loadContacts, boolean loadParent) {
		CriteriaBuilder qb = getSession().getCriteriaBuilder();

		// Create CriteriaQuery
		CriteriaQuery<Task> criteria = qb.createQuery(Task.class);
		Root<Task> root = criteria.from(Task.class);
		criteria.select(root);

		Join<Task, Sample> sampleQuery = root.join("samples", JoinType.LEFT);
		Join<Sample, Block> blockQuery = sampleQuery.join("blocks", JoinType.LEFT);
		Join<Block, Slide> slideQuery = blockQuery.join("slides", JoinType.LEFT);

		return find(criteria, root,
				Arrays.asList(qb.and(qb.like(root.get("taskID"), taskID),
						qb.equal(slideQuery.get("uniqueIDinTask"), uniqueSlideIDInBlock))),
				loadCouncils, loadDiangoses, loadPDFs, loadContacts, loadParent);
	}

	public Optional<Task> findOptinalByLastID(Calendar ofYear, boolean loadCouncils, boolean loadDiangoses, boolean loadPDFs,
			boolean loadContacts, boolean loadParent) {

		String currentYear = Integer.toString(TimeUtil.getYearAsInt(ofYear) - 2000) + "%";

		// Create CriteriaBuilder
		CriteriaBuilder qb = getSession().getCriteriaBuilder();

		// Create CriteriaQuery
		CriteriaQuery<Task> criteria = qb.createQuery(Task.class);
		Root<Task> root = criteria.from(Task.class);

		Subquery<Task> subquery = criteria.subquery(Task.class);
		Root<Task> subTaskRoot = subquery.from(Task.class);
		subquery.where(qb.like(subTaskRoot.get("taskID"), currentYear));
		subquery.select(qb.max((Expression) subTaskRoot.get("taskID")));

		return find(criteria, root, Arrays.asList(qb.equal(root.get("taskID"), subquery)), loadCouncils, loadDiangoses,
				loadPDFs, loadContacts, loadParent);
	}

	public List<Task> findByCriteria(WorklistSearchExtended worklistSearchExtended, boolean initParent) {
		CriteriaBuilder qb = getSession().getCriteriaBuilder();

		// Create CriteriaQuery
		CriteriaQuery<Task> criteria = qb.createQuery(Task.class);
		Root<Task> root = criteria.from(Task.class);
		criteria.select(root);

		if (initParent)
			root.fetch("parent", JoinType.LEFT);

		Join<Task, Sample> sampleQuery = root.join("samples", JoinType.LEFT);
		Join<Sample, Block> blockQuery = sampleQuery.join("blocks", JoinType.LEFT);
		Join<Block, Slide> slideQuery = blockQuery.join("slides", JoinType.LEFT);
		Join<Slide, StainingPrototype> prototypeQuery = slideQuery.join("slidePrototype", JoinType.LEFT);

		Join<Task, DiagnosisRevision> diagnosisRevisionQuery = root.join("diagnosisRevisions", JoinType.LEFT);
		Join<DiagnosisRevision, Diagnosis> diagnosesQuery = diagnosisRevisionQuery.join("diagnoses", JoinType.LEFT);
		Join<DiagnosisRevision, Signature> signatureOneQuery = diagnosisRevisionQuery.join("signatureOne",
				JoinType.LEFT);
		Join<Signature, Physician> signatureOnePhysicianQuery = signatureOneQuery.join("physician", JoinType.LEFT);
		Join<DiagnosisRevision, Signature> signatureTwoQuery = diagnosisRevisionQuery.join("signatureTwo",
				JoinType.LEFT);
		Join<Signature, Physician> signatureTwoPhysicianQuery = signatureTwoQuery.join("physician", JoinType.LEFT);

		Join<ReportIntent, Task> contactQuery = root.join("contacts", JoinType.LEFT);
		Join<Person, ReportIntent> personContactQuery = contactQuery.join("person", JoinType.LEFT);

		List<Predicate> predicates = new ArrayList<Predicate>();

		// searching for material
		if (HistoUtil.isNotNullOrEmpty(worklistSearchExtended.getMaterial())) {
			predicates.add(qb.like(qb.lower(sampleQuery.get("material")),
					"%" + worklistSearchExtended.getMaterial().toLowerCase() + "%"));

			logger.debug("Selecting material " + worklistSearchExtended.getMaterial());
		}

		// getting surgeon
		if (HistoUtil.isNotNullOrEmpty(worklistSearchExtended.getSurgeons())) {
			Expression<Long> exp = personContactQuery.get("id");
			predicates
					.add(qb.and(
							exp.in(Arrays.asList(worklistSearchExtended.getSurgeons()).stream()
									.map(p -> p.getPerson().getId()).collect(Collectors.toList())),
							qb.equal(contactQuery.get("role"), ContactRole.SURGEON)));
			logger.debug("Selecting surgeon");
		}

		// getting signature
		if (HistoUtil.isNotNullOrEmpty(worklistSearchExtended.getSignature())) {
			Expression<Long> expphysicianOne = signatureOnePhysicianQuery.get("id");
			Predicate physicianOne = expphysicianOne.in(Arrays.asList(worklistSearchExtended.getSignature()).stream()
					.map(p -> p.getId()).collect(Collectors.toList()));

			Expression<Long> expphysicianTwo = signatureTwoPhysicianQuery.get("id");
			Predicate physicianTwo = expphysicianTwo.in(Arrays.asList(worklistSearchExtended.getSignature()).stream()
					.map(p -> p.getId()).collect(Collectors.toList()));

			predicates.add(qb.or(physicianOne, physicianTwo));

			logger.debug("Selecting signature");
		}

		// getting history
		if (HistoUtil.isNotNullOrEmpty(worklistSearchExtended.getCaseHistory())) {
			predicates.add(qb.like(qb.lower(root.get("caseHistory")),
					"%" + worklistSearchExtended.getCaseHistory().toLowerCase() + "%"));

			logger.debug("Selecting history");
		}

		// getting diagnosis text
		if (HistoUtil.isNotNullOrEmpty(worklistSearchExtended.getDiagnosisText())) {
			predicates.add(qb.like(qb.lower(diagnosisRevisionQuery.get("text")),
					"%" + worklistSearchExtended.getDiagnosisText().toLowerCase() + "%"));

			logger.debug("Selecting diagnosis text");
		}

		// getting diagnosis
		if (HistoUtil.isNotNullOrEmpty(worklistSearchExtended.getDiagnosis())) {
			predicates.add(qb.like(qb.lower(diagnosesQuery.get("diagnosis")),
					"%" + worklistSearchExtended.getDiagnosis().toLowerCase() + "%"));

			logger.debug("Selecting diagnosis");
		}

		// checking malign, 0 = not selected, 1 = true, 2 = false
		if (!worklistSearchExtended.getMalign().equals("0")) {
			predicates.add(qb.equal(diagnosesQuery.get("malign"),
					worklistSearchExtended.getMalign().equals("1") ? true : false));

			logger.debug("Selecting malign");
		}

		// getting eye
		if (worklistSearchExtended.getEye() != Eye.UNKNOWN) {
			predicates.add(qb.equal(root.get("eye"), worklistSearchExtended.getEye()));

			logger.debug("Selecting eye");
		}

		// getting ward
		if (HistoUtil.isNotNullOrEmpty(worklistSearchExtended.getWard())) {
			predicates.add(qb.like(qb.lower(diagnosesQuery.get("ward")),
					"%" + worklistSearchExtended.getWard().toLowerCase() + "%"));

			logger.debug("Selecting ward");
		}

		if (worklistSearchExtended.getStainings() != null && !worklistSearchExtended.getStainings().isEmpty()) {
			Expression<Long> prototypeID = prototypeQuery.get("id");
			Predicate stainings = prototypeID.in(
					worklistSearchExtended.getStainings().stream().map(p -> p.getId()).collect(Collectors.toList()));

			predicates.add(stainings);
		}

		criteria.where(qb.and(predicates.toArray(new Predicate[predicates.size()])));

		criteria.distinct(true);

		List<Task> tasks = getSession().createQuery(criteria).getResultList();

		return tasks;
	}

	public Optional<Task> find(CriteriaQuery<Task> criteria, Root<Task> root, List<Predicate> predicates,
			boolean loadCouncils, boolean loadDiangoses, boolean loadPDFs, boolean loadContacts, boolean loadParent) {
		criteria.select(root);

		if (loadCouncils)
			root.fetch(Task_.councils, JoinType.LEFT);

		if (loadDiangoses)
			root.fetch(Task_.diagnosisRevisions, JoinType.LEFT);

		if (loadContacts)
			root.fetch(Task_.contacts, JoinType.LEFT);

		if (loadPDFs)
			root.fetch(Task_.attachedPdfs, JoinType.LEFT);

		criteria.where(getCriteriaBuilder().and(predicates.toArray(new Predicate[predicates.size()])));
		criteria.distinct(true);

		List<Task> groups = getSession().createQuery(criteria).getResultList();

		Optional<Task> task = Optional.ofNullable(groups.size() > 0 ? groups.get(0) : null);

		if (loadParent && task.isPresent()) {
			Hibernate.initialize(task.get().getParent().getTasks());
			Hibernate.initialize(task.get().getParent().getAttachedPdfs());
		}

		return task;
	}
}

File: PathoFroh/src/main/kotlin/com/patho/main/repository/custom/impl/TaskRepositoryImpl.kt
package com.patho.main.repository.custom.impl

import com.patho.main.model.Physician
import com.patho.main.model.Signature
import com.patho.main.model.Signature_
import com.patho.main.model.StainingPrototype
import com.patho.main.model.patient.*
import com.patho.main.model.patient.notification.ReportIntent
import com.patho.main.model.patient.notification.ReportIntent_
import com.patho.main.model.person.Person
import com.patho.main.model.person.Person_
import com.patho.main.repository.custom.TaskRepositoryCustom
import com.patho.main.repository.service.impl.AbstractRepositoryCustom
import com.patho.main.util.exceptions.TaskNotFoundException
import com.patho.main.util.search.settings.ExtendedSearch
import org.hibernate.Hibernate
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import javax.persistence.criteria.*


@Service
open class TaskRepositoryImpl : AbstractRepositoryCustom(), TaskRepositoryCustom {

    @Transactional
    override fun findByID(task: Task, loadCouncils: Boolean, loadDiagnoses: Boolean,
                          loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): Task {
        return findByID(task.id, loadCouncils, loadDiagnoses, loadPDFs, loadContacts, loadParent)
    }

    @Transactional
    override fun findByID(id: Long, loadCouncils: Boolean, loadDiagnoses: Boolean,
                          loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): Task {
        val criteria = criteriaBuilder.createQuery(Task::class.java)
        val root = criteria.from(Task::class.java)

        val tasks = findAll(criteria, root, mutableListOf(criteriaBuilder.equal(root.get(Task_.id), id)), loadCouncils, loadDiagnoses,
                loadPDFs, loadContacts, loadParent)

        return tasks.firstOrNull() ?: throw TaskNotFoundException()
    }

    @Transactional
    override fun findByTaskID(taskID: String, loadCouncils: Boolean, loadDiagnoses: Boolean,
                              loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): Task {
        val criteria = criteriaBuilder.createQuery(Task::class.java)
        val root = criteria.from(Task::class.java)
        criteria.select(root)

        val tasks = findAll(criteria, root, mutableListOf(criteriaBuilder.like(root.get(Task_.taskID), taskID)), loadCouncils, loadDiagnoses,
                loadPDFs, loadContacts, loadParent)

        return tasks.firstOrNull() ?: throw TaskNotFoundException()
    }

    @Transactional
    override fun findBySlideID(taskID: String, uniqueSlideIDInBlock: Int, loadCouncils: Boolean,
                               loadDiagnoses: Boolean, loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): Task {
        val criteria = criteriaBuilder.createQuery(Task::class.java)
        val root = criteria.from(Task::class.java)
        criteria.select(root)

        val sampleQuery = root.join<Task, Sample>(Task_.samples.name, JoinType.LEFT)
        val blockQuery = sampleQuery.join<Sample, Block>(Sample_.blocks.name, JoinType.LEFT)
        val slideQuery = blockQuery.join<Block, Slide>(Block_.slides.name, JoinType.LEFT)

        val tasks = findAll(criteria, root,
                mutableListOf(criteriaBuilder.and(criteriaBuilder.like(root.get(Task_.taskID), taskID),
                        criteriaBuilder.equal(slideQuery.get(Slide_.uniqueIDinTask), uniqueSlideIDInBlock))),
                loadCouncils, loadDiagnoses, loadPDFs, loadContacts, loadParent)

        return tasks.firstOrNull() ?: throw TaskNotFoundException()
    }

    @Transactional
    override fun findByLastID(ofYear: LocalDate, loadCouncils: Boolean, loadDiagnoses: Boolean, loadPDFs: Boolean,
                              loadContacts: Boolean, loadParent: Boolean): Task {

        val currentYear = (ofYear.year - 2000).toString() + "%"

        // Create CriteriaQuery
        val criteria = criteriaBuilder.createQuery(Task::class.java)
        val root = criteria.from(Task::class.java)

        val subquery = criteria.subquery(Number::class.java)
        val subTaskRoot = subquery.from(Task::class.java)
        subquery.where(criteriaBuilder.like(subTaskRoot.get(Task_.taskID), currentYear))
        subquery.select(criteriaBuilder.max(subTaskRoot.get(Task_.taskID) as Expression<Number>))

        val tasks = findAll(criteria, root, mutableListOf(criteriaBuilder.equal(root.get(Task_.taskID), subquery)), loadCouncils, loadDiagnoses,
                loadPDFs, loadContacts, loadParent)

        return tasks.firstOrNull() ?: throw TaskNotFoundException()
    }

    @Transactional
    override fun findByExtendedSearchSettings(extendedSearch: ExtendedSearch, loadCouncils: Boolean,
                                              loadDiagnoses: Boolean, loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): List<Task> {
        val criteria = criteriaBuilder.createQuery(Task::class.java)
        val root = criteria.from(Task::class.java)
        criteria.select(root)

        val patientQuery: Join<Task, Patient> = root.join(Task_.parent.name, JoinType.LEFT)
        val personPatientQuery: Join<Patient, Person> = patientQuery.join(Patient_.person.name, JoinType.LEFT)

        val sampleQuery: Join<Task, Sample> = root.join(Task_.samples.name, JoinType.LEFT)
        val blockQuery: Join<Sample, Block> = sampleQuery.join(Sample_.blocks.name, JoinType.LEFT)
        val slideQuery: Join<Block, Slide> = blockQuery.join(Block_.slides.name, JoinType.LEFT)
        val prototypeQuery: Join<Slide, StainingPrototype> = slideQuery.join(Slide_.slidePrototype.name, JoinType.LEFT)

        val diagnosisRevisionQuery: Join<Task, DiagnosisRevision> = root.join(Task_.diagnosisRevisions.name, JoinType.LEFT)
        val diagnosesQuery: Join<DiagnosisRevision, Diagnosis> = diagnosisRevisionQuery.join(DiagnosisRevision_.diagnoses.name, JoinType.LEFT)
        val signatureOneQuery: Join<DiagnosisRevision, Signature> = diagnosisRevisionQuery.join(DiagnosisRevision_.signatureOne.name, JoinType.LEFT)
        val signatureOnePhysicianQuery: Join<Signature, Physician> = signatureOneQuery.join(Signature_.physician.name, JoinType.LEFT)
        val signatureTwoQuery: Join<DiagnosisRevision, Signature> = diagnosisRevisionQuery.join(DiagnosisRevision_.signatureTwo.name,
                JoinType.LEFT)
        val signatureTwoPhysicianQuery: Join<Signature, Physician> = signatureTwoQuery.join(Signature_.physician.name, JoinType.LEFT)

        val contactQuery: Join<ReportIntent, Task> = root.join(Task_.contacts.name, JoinType.LEFT)
        val personContactQuery: Join<Person, ReportIntent> = contactQuery.join(ReportIntent_.person.name, JoinType.LEFT)

        val predicates = mutableListOf<Predicate>()

        // patient name
        if (extendedSearch.isUsePatientName) {
            predicates.add(criteriaBuilder.like(criteriaBuilder.lower(personPatientQuery.get(Person_.lastName)),
                    "%${extendedSearch.patientName}%"))
            logger.debug("Search for task -> patient -> person.name ${extendedSearch.patientName}")
        }

        // surname
        if (extendedSearch.isUsePatientSurname) {
            predicates.add(criteriaBuilder.like(criteriaBuilder.lower(personPatientQuery.get(Person_.firstName)),
                    "%${extendedSearch.patientSurname}%"))
            logger.debug("Search for task -> patient -> person.firstname ${extendedSearch.patientSurname}")
        }

        // birthday from
        if (extendedSearch.isUsePatientBirthdayFrom && !extendedSearch.isUsePatientBirthdayTo) {
            predicates.add(criteriaBuilder.equal(personPatientQuery.get(Person_.birthday),
                    extendedSearch.patientBirthdayFrom))
            logger.debug("Search for task -> patient -> person.birthday ${extendedSearch.patientBirthdayFrom}")
        }

        // birthday from to
        if (extendedSearch.isUsePatientBirthdayFrom && extendedSearch.isUsePatientBirthdayTo) {
            predicates.add(criteriaBuilder.between(personPatientQuery.get(Person_.birthday),
                    extendedSearch.patientBirthdayFrom, extendedSearch.patientBirthdayTo))
            logger.debug("Search for task -> patient -> person.birthday ${extendedSearch.patientBirthdayFrom} to${extendedSearch.patientBirthdayTo} ")
        }

        // material
        if (extendedSearch.isUseMaterial) {

            val materials = mutableListOf<Predicate>()

            extendedSearch.material?.forEach {
                materials.add(criteriaBuilder.like(criteriaBuilder.lower(sampleQuery.get(Sample_.material)),
                        "%$it%"))
                logger.debug("Search for task -> sample -> material $it")
            }

            if (materials.size == 1)
                predicates.addAll(materials)
            else if (materials.size > 1) {
                val or = criteriaBuilder.or(*materials.toTypedArray())
                predicates.add(or)
            }
        }

        // case history
        if (extendedSearch.isUseCaseHistory) {

            val history = mutableListOf<Predicate>()

            extendedSearch.caseHistory?.forEach {
                history.add(criteriaBuilder.like(criteriaBuilder.lower(root.get(Task_.caseHistory)),
                        "%$it%"))
                logger.debug("Search for task -> caseHistory $it")
            }

            if (history.size == 1)
                predicates.addAll(history)
            else if (history.size > 1) {
                val or = criteriaBuilder.or(*history.toTypedArray())
                predicates.add(or)
            }
        }

        return findAll(criteria, root, predicates, loadCouncils, loadDiagnoses, loadPDFs, loadContacts, loadParent)
    }

    @Transactional
    override fun findAll(criteria: CriteriaQuery<Task>, root: Root<Task>, predicates: MutableList<Predicate>,
                         loadCouncils: Boolean, loadDiagnoses: Boolean, loadPDFs: Boolean, loadContacts: Boolean, loadParent: Boolean): List<Task> {
        criteria.select(root)

        val graph = session.createEntityGraph(Task::class.java)

        if (loadCouncils)
            graph.addAttributeNodes(Task_.councils.name)

        if (loadDiagnoses)
            graph.addAttributeNodes(Task_.diagnosisRevisions.name)

        if (loadContacts)
            graph.addAttributeNodes(Task_.contacts.name)

        if (loadPDFs)
            graph.addAttributeNodes(Task_.attachedPdfs.name)

        if (loadParent) {
            val subgraph = graph.addSubGraph(Task_.parent.name, Patient::class.java)
            subgraph.addAttributeNodes(Patient_.tasks.name)
            subgraph.addAttributeNodes(Patient_.attachedPdfs.name)
        }

        criteria.where(criteriaBuilder.and(*predicates.toTypedArray()))
        criteria.distinct(true)

        val tasks = session.createQuery<Task>(criteria).setHint("javax.persistence.loadgraph", graph).resultList;

//
//        tasks.forEach {
//            Hibernate.initialize(it.parent!!.tasks)
//            Hibernate.initialize(it.parent!!.attachedPdfs)
//        }

        return tasks
    }

}